src/tools/compiletest/src/load_cfg/itemlist.rs

```diff
@@ -0,0 +1,733 @@
+    RemapSrcBase(bool),
+
+    /* Key-value directivesthat can be set with `foo-bar: baz qux` */
+    ErrorPattern(&'src str),
+    RegexErrorPattern(&'src str),
+    CompileFlags(&'src str),
+    Edition(&'src str),
+    RunFlags(&'src str),
+    PrettyMode(&'src str),
+    AuxBin(&'src str),
+    AuxBuild(&'src str),
+    AuxCrate(&'src str),
+    AuxCodegenBackend(&'src str),
+    ExecEnv(&'src str),
+    UnsetExecEnv(&'src str),
+    RustcEnv(&'src str),
+    UnsetRustcEnv(&'src str),
+    ForbidOutput(&'src str),
+    FailureStatus(&'src str),
+    AssemblyOutput(&'src str),
+    TestMirPass(&'src str),
+    LlvmCovFlags(&'src str),
+    FilecheckFlags(&'src str),
+    /// The revisions list
+    Revisions(&'src str),
+
+    /* Directives that are single keys but have variable suffixes */
+    /// Ignore something specific about this test.
+    ///
+    /// `//@ ignore-x86`
+    Ignore {
+        what: &'src str,
+    },
+    /// Components needed by the test
+    ///
+    /// `//@ needs-rust-lld`
+    Needs {
+        what: &'src str,
+    },
+    /// Run only if conditions are met
+    ///
+    /// `//@ only-linux`
+    Only {
+        what: &'src str,
+    },
+    /// Normalizations make transformations in output files before checking
+    ///
+    /// `//@ normalize-stdout: foo -> bar`
+    /// `//@ normalize-stderr-32bit: bar -> baz`
+    /// `//@ normalize-stderr-test: "h[[:xdigit:]]{16}" -> "h[HASH]"`
+    Normalize {
+        what: &'src str,
+    },
+
+    /* Specialized patterns */
+    /// Directives that apply to a single revision. Don't process this here; just extract the
+    /// key and value. The content will be reprocessed as a new item.
+    ///
+    /// Note that this does not yet recurse.
+    ///
+    /// `//@[revname] directive`
+    RevisionSpecificItems {
+        revs: &'src str,
+        content: &'src str,
+    },
+
+    /// Once we have parsed the entire file, we re-parse `RevisionSpecificItems` into
+    /// `RevisionSpecificExpanded`.
+    RevisionSpecificExpanded {
+        revs: &'src str,
+        content: Box<Self>,
+    },
+
+    /* Non-header things that we consume. */
+    UiDirective {
+        /// If specified, a list of which revisions this should check for
+        revisions: Option<&'src str>,
+        /// `^`, `^^`, `|`, `~`, etc
+        adjust: Option<&'src str>,
+        /// Directives for filecheck, including `ERROR`, `WARN`, `HELP`, etc
+        content: &'src str,
+    },
+    /// Directives intended for filecheck.
+    ///
+    /// `// CHECK:`, `// CHECK-NEXT`, ...
+    FileCheckDirective {
+        directive: &'src str,
+        content: Option<&'src str>,
+    },
+}
+
+impl<'src> ItemVal<'src> {
+    pub const fn to_item(self, line: usize, col: usize) -> Item<'src> {
+        Item { val: self, pos: LineCol::new(line, col), used: Cell::new(false) }
+    }
+
+    /// True if this is an `//@` directive, i.e. something in a header`
+    pub fn is_header_directive(&self) -> bool {
```

Comment:
Discussion: compiletest mixes the terms directives, headers and cfgs, but AFAICT they are trying to say the same thing. Headers do not, in fact, have to appear in the head of the file (that was a historical restriction AFAIK).

Reply:
I have just been using "header directives" to refer to `//@` things, "UI directives" to refer to `//~ ERROR` things, and "filecheck directives" to refer to `// CHECK:` things. But I agree this isn't as clear as it could be (nor consistent across documentation) - do you have preferred names? "Config directive" instead of "header directives" maybe.

(I actually currently enforce that config directives are above UI or filecheck directives as kind of a tidy thing, since having config directives later in the file seems easy to miss. I did see on Zulip though that sometimes you want to be able to append directives - maybe we can leave the check as-is but add a `--no-check-directive-order` flag to opt out. Seems like possible overkill, not sure if that is worth it).